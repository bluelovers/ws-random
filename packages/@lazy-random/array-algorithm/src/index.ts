
export * from './swapAlgorithm';
export * from './array_rebase';
