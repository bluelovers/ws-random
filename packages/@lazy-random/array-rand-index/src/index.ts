import { _MathRandom } from '@lazy-random/original-math-random';

export function arrayRandIndexByLength(len: number, ...argv: any[])
{
	return Math.floor(_MathRandom() * len)
}

export function arrayRandIndex(array: any[], ...argv: any[])
{
	return arrayRandIndexByLength(array.length)
}

export default arrayRandIndex
