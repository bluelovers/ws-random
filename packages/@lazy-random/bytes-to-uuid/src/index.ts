import { BYTE_TO_HEX_TO_LOWER_CASE } from '@lazy-random/shared-lib';

export function _createBytesToUuidFn(bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	return (buf: ArrayLike<number>, offset?: number) =>
	{
		let i = offset || 0;
		// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
		return ([
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
		]).join('');
	}
}

/**
 * @see https://github.com/kelektiv/node-uuid/blob/master/lib/bytesToUuid.js
 */
export function bytesToUuid(buf: ArrayLike<number>, offset?: number, bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	let i = offset || 0;
	// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
	return ([
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
	]).join('');
}

export default bytesToUuid
