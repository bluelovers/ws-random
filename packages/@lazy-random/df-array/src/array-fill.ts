import { dfUniformByte, dfUniformFloat, dfUniformInt } from '@lazy-random/df-uniform';
import { expect } from '@lazy-random/expect';
import { IArrayInput02 } from '@lazy-random/shared-lib';
import { isUnset } from '@lazy-random/shared-lib';
import { IRNGLike } from '@lazy-random/rng-abstract';

export function dfArrayFill(random: IRNGLike, min?: number, max?: number, float?: boolean)
{
	let fn: () => number;

	{
		let min_unset = isUnset(min);
		let max_unset = isUnset(max);

		if (max_unset && min_unset)
		{
			fn = dfUniformByte(random);
		}
		else if (float)
		{
			fn = dfUniformFloat(random, min, max);
		}
		else
		{
			fn = dfUniformInt(random, min, max);
		}

		min = void 0;
		max = void 0;
	}

	expect(fn).function();

	return <T extends IArrayInput02<number>>(arr: T) =>
	{
		let i = arr.length;
		while (i--)
		{
			arr[i] = fn();
		}
		return arr
	}
}

export default dfArrayFill
