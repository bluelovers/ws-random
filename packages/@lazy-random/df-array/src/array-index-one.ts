import { IRNGLike } from '@lazy-random/rng-abstract-core';
import { ITSArrayListMaybeReadonly } from 'ts-type/lib/type/base';
import { _handleStartEnd } from './util/options';
import { int } from '@lazy-random/util-distributions';

/**
 * return index number form array
 */
export function dfArrayIndexOne<T extends ITSArrayListMaybeReadonly<unknown>>(random: IRNGLike,
	arr: T,
	start: number = 0,
	end?: number
)
{
	({
		start,
		end,
	} = _handleStartEnd(arr, start, end));

	return () =>
	{
		return int(random, start, end)
	}
}

export default dfArrayIndexOne
