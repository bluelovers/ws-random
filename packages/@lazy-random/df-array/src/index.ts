
export { dfArrayIndex } from './array-index'
export { dfArrayShuffle } from './array-shuffle'

export { dfArrayUnique } from './array-unique'
export { dfArrayFill } from './array-fill'

