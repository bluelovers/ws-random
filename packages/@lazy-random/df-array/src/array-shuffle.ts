import { ITSArrayLikeWriteable } from 'ts-type/lib/generic';
import { TypedArray } from 'typedarray-dts';
import { swapAlgorithm2 } from '@lazy-random/array-algorithm';
import { randIndex as _randIndex } from '@lazy-random/util-distributions';
import { IRNGLike } from '@lazy-random/rng-abstract';

export function dfArrayShuffle<T extends ITSArrayLikeWriteable<any> | TypedArray | Buffer>(random: IRNGLike, arr: T, overwrite?: boolean): () => T
{
	const randIndex = (len: number) =>
	{
		return _randIndex(random, len)
	};

	if (!overwrite)
	{
		let cloneArrayLike: (arr: T) => T;

		if (Buffer.isBuffer(arr))
		{
			// @ts-ignore
			cloneArrayLike = (arr) =>
			{
				// @ts-ignore
				return Buffer.from(arr)
			};
		}
		else
		{
			cloneArrayLike = (arr) =>
			{
				// @ts-ignore
				return arr.slice()
			};
		}

		return (): T =>
		{
			return swapAlgorithm2(cloneArrayLike(arr), true, randIndex)
		}
	}

	return (): T =>
	{
		return swapAlgorithm2(arr, true, randIndex)
	}
}

dfArrayShuffle.memoizable = false;

export default dfArrayShuffle;


