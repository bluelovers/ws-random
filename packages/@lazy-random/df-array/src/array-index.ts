import { expect } from '@lazy-random/expect';
import { IRNGLike } from '@lazy-random/rng-abstract';
import { _handleStartEnd } from './util/options';
import { dfArrayIndexOne } from './array-index-one';
import { ITSArrayListMaybeReadonly } from 'ts-type/lib/type/base';

/**
 * return index list form array
 */
export function dfArrayIndex<T extends ITSArrayListMaybeReadonly<unknown>>(random: IRNGLike, arr: T, size: number = 1, start: number = 0, end?: number)
{
	expect(size).integer.gt(0);

	let len: number;

	({
		start,
		end,
		len,
	} = _handleStartEnd(arr, start, end));

	let size_runtime = Math.max(Math.min(end - start, len, size), 0);

	expect(size_runtime).gte(size).gt(0)

	const fn = dfArrayIndexOne(random, arr, start, end);

	return () =>
	{
		let ids: number[] = [];
		let prev: number;

		LABEL_TOP: do
		{
			let i = fn()

			if (prev === i || ids.includes(i))
			{
				continue LABEL_TOP;
			}

			ids.push(prev = i);
			--size_runtime;
		}
		while (size_runtime > 0);

		return ids;
	}
}

export default dfArrayIndex
