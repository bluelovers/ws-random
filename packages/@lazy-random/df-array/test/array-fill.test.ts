import { newRngSeedRandom } from '@lazy-random/util-test';
import { dfArrayFill } from '../src/index';

describe(`dfArrayFill`, () =>
{
	const fn = dfArrayFill(newRngSeedRandom());

	const tests = [
		['Array', new Array(10)],
		['Uint8Array', new Uint8Array(10)],
		['Buffer', Buffer.alloc(10)],
	] as const;

	tests.forEach(function (arr)
	{
		it(`${arr[0]}`, () =>
		{
			let ret = fn(arr[1]);

			expect(ret).toHaveLength(10);
			expect(ret).toMatchSnapshot()
		});
	});
});
