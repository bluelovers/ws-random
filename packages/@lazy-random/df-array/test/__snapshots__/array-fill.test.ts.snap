// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dfArrayFill Array 1`] = `
Array [
  241,
  236,
  111,
  63,
  1,
  106,
  184,
  72,
  141,
  19,
]
`;

exports[`dfArrayFill Buffer 1`] = `
Object {
  "data": Array [
    239,
    252,
    20,
    187,
    246,
    119,
    163,
    7,
    172,
    97,
  ],
  "type": "Buffer",
}
`;

exports[`dfArrayFill Uint8Array 1`] = `
Uint8Array [
  238,
  80,
  56,
  164,
  182,
  235,
  19,
  3,
  187,
  246,
]
`;
